### PR TITLE
buildsystem: more optimisations

### PR DESCRIPTION
**Pull request by MilhouseVH**
Another small round of buildsystem optimisations, this time eliminating about 22 sub-process/forks per call to `config/options <packagename>` (which is executed at least 3 times per package).

Using the same test script from [#2072](https://github.com/LibreELEC/LibreELEC.tv/pull/2072#issuecomment-334957175), the following results were obtained (AMD FX-8350).

Before:
```
ELAPSED: 142 seconds, 557 packages x3 (LOOP #1)
ELAPSED: 144 seconds, 557 packages x3 (LOOP #2)
ELAPSED: 145 seconds, 557 packages x3 (LOOP #3)
ELAPSED: 146 seconds, 557 packages x3 (LOOP #4)
ELAPSED: 145 seconds, 557 packages x3 (LOOP #5)
```

After:
```
ELAPSED: 76 seconds, 557 packages x3 (LOOP #1)
ELAPSED: 78 seconds, 557 packages x3 (LOOP #2)
ELAPSED: 77 seconds, 557 packages x3 (LOOP #3)
ELAPSED: 78 seconds, 557 packages x3 (LOOP #4)
ELAPSED: 79 seconds, 557 packages x3 (LOOP #5)
```

So, a worthwhile improvement.

The general idea is to avoid using external binaries when shell built-ins can be used instead, and to export variables that are "expensive" to acquire and that don't change over the lifetime of the build (eg. `HOST_NAME`).

This has been tested with clean RPi/RPi2/Generic builds.

### CODE REVIEW

File: scripts/build [hunk 1]
```diff
@@ -31,15 +31,18 @@ if [ ! -f $PKG_DIR/package.mk ]; then
...
+PKG_MAKEINSTALL_OPTS=""
+
+if [ "${1//:/}" != "${1}" ]; then
+  PACKAGE_NAME="${1%:*}"
+  TARGET="${1#*:}"
+else
+  PACKAGE_NAME=$1
+  TARGET=
+fi
+[ -z "$TARGET" ] && TARGET="target"
```
[the remark this answers is not in the document]
Comment (MilhouseVH) [reply scripts/build]: I did consider something like that, but the thing I don't like is hard-coding the default value twice, which just "feels" wrong/ugly, and the time saved from doing that is likely to be almost impossible to measure.

File: config/optimize [hunk 2]
```diff
@@ -40,7 +40,9 @@ HOST_CXXFLAGS="$HOST_CXXFLAGS -Wno-format-security"
 
 # add distro specific library dirs
   # ubuntu/debian specific "multiarch support"
-    FAMILY_TRIPLET=$(echo $HOST_NAME | sed -e "s,$(uname -m),$(uname -i),")
+    [ -z "${MACHINE_HARDWARE_NAME}" ] && MACHINE_HARDWARE_NAME="$(uname -m)"
+    [ -z "${MACHINE_HARDWARE_PLATFORM}" ] && MACHINE_HARDWARE_PLATFORM="$(uname -i)"
```
Comment (MilhouseVH) [inline config/optimize]: I've missed a couple of `export`s here...

Comment (MilhouseVH) [reply config/optimize]: Now added.

File: config/optimize [hunk 3]
```diff
@@ -49,7 +51,7 @@ HOST_CXXFLAGS="$HOST_CXXFLAGS -Wno-format-security"
     fi
 
   # default multiarch support
-    case "`uname -m`" in
+    case "${MACHINE_HARDWARE_NAME}" in
```
Comment (MilhouseVH) [inline config/optimize]: In fact, once we have determined `HOST_LIBDIR` it should be possible to export it avoid all of this processing in subsequent passes. Will update.

**Comment by MilhouseVH** [comment]
Just doing some more clean builds and got a weird error, need to investigate so don't merge this yet...

**Comment by MilhouseVH** [comment]
I think this is OK. Phew.

Last night I left 9 concurrent builds compiling overnight (RPi with 4.13.y, RPi with 4.13.y and debug and RPi with 4.14.y, and the same for RPi2 and Generic) and the OOM process killer killed the first two Generic builds while they were building binutils:
```
[520055.426294] Out of memory: Kill process 28895 (cc1plus) score 12 or sacrifice child
[520055.426298] Killed process 28895 (cc1plus) total-vm:440212kB, anon-rss:91500kB, file-rss:11180kB
[520102.243690] bash invoked oom-killer: gfp_mask=0x26000c0, order=2, oom_score_adj=0
```
which resulted in:
```
mv -f .deps/aarch64-reloc-property.Tpo .deps/aarch64-reloc-property.Po
/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/toolchain/bin/host-g++ -DHAVE_CONFIG_H -I. -I/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/binutils-2.29/gold  -I/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/binutils-2.29/gold -I/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/binutils-2.29/gold/../include -I/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/binutils-2.29/gold/../elfcpp -DLOCALEDIR="\"/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/toolchain/share/locale\"" -DBINDIR="\"/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/toolchain/bin\"" -DTOOLBINDIR="\"/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/toolchain/x86_64-libreelec-linux-gnu/bin\"" -DTOOLLIBDIR="\"/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/toolchain/x86_64-libreelec-linux-gnu/lib\""   -W -Wall    -Wstack-usage=262144 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -frandom-seed=incremental-dump.o -I/home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/binutils-2.29/gold/../zlib -g -O2 -MT incremental-dump.o -MD -MP -MF .deps/incremental-dump.Tpo -c -o incremental-dump.o /home/neil/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.0-devel/binutils-2.29/gold/incremental-dump.cc
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-5/README.Bugs> for instructions.
make[5]: *** [Makefile:918: mips.o] Error 4
make[5]: *** Waiting for unfinished jobs....
```
So this PR is OK. I just need more RAM/swap. :)

**Comment by MilhouseVH** [comment]
I've added a `listcontains` function that is quite efficient, and might come in handy.

**Comment by MilhouseVH** [comment]
And just for the sake of it... a simple test program:
```
#!/bin/bash

listcontains1() {
  [[ $1 =~ (^|[[:space:]])$2($|[[:space:]]) ]] && return 0 || return 1
}

listcontains2() {
  local pkg
  for pkg in $1; do
    [ "${pkg}" = "${2}" ] && return 0
  done
  return 1
}

gettime() {
  echo "$(date +%s.%N)"
}

getdelta() {
  python -c "print('%6.3f' % float(${2} - ${1}))"
}

LOOP=${LOOP:-10000}
LISTSIZE=${LISTSIZE:-20}

LIST="start"
for i in $(seq ${LISTSIZE}); do LIST+=" abc 123 def 456 ghi 789"; done
LIST+=" middle"
for i in $(seq ${LISTSIZE}); do LIST+=" abc 123 def 456 ghi 789"; done
LIST+=" end"

c=0
s=$(gettime)
while [ $c -lt $LOOP ]; do
  c=$((c+1))
  listcontains1 "${LIST}" "${1}" && r= || r="not "
done
e=$(gettime)
echo "ELAPSED (regex   ): $(getdelta ${s} ${e}) seconds (\"${1}\" is ${r}in list)"

c=0
s=$(gettime)
while [ $c -lt $LOOP ]; do
  c=$((c+1))
  listcontains2 "${LIST}" "${1}" && r= || r="not "
done
e=$(gettime)
echo "ELAPSED (for loop): $(getdelta ${s} ${e}) seconds (\"${1}\" is ${r}in list)"
```
Run as:
```
LOOP=100000 LISTSIZE=5 /tmp/listtest start
LOOP=100000 LISTSIZE=5 /tmp/listtest middle
LOOP=100000 LISTSIZE=5 /tmp/listtest end
LOOP=100000 LISTSIZE=5 /tmp/listtest notfound
```
Results on FX-8350 (times in seconds):

| string | result | regex | for loop |
| :---- | :----: | :----: | :----: |
| `start`  | in list | 8.9 | 7.2 |
| `middle` | in list | 9.7 | 27.3 |
| `end`   | in list | 9.9 | 46.9 |
| `notfound` | not in list | 9.6 | 48.6 |

**Comment by lrusak** [comment]
Good to go?

**Comment by MilhouseVH** [comment]
Yes

**Pull request merged**